### PR TITLE
internal: add a FIFOList to only store n recent inserts

### PR DESCRIPTION
**Pull request by jhchabran**
In the context of https://github.com/sourcegraph/sourcegraph/pull/45063 I needed to be able to store elements in a list, while making sure it never grows over a certain size. 

While this could be handled by manually deleting entries and using the existing `rcache.Cache`, it has the shortcoming of requiring to list all keys to do this, which can produce unexpected results if not careful with how many keys has been inserted. 

So instead, I added a new type for dealing with these, which I've named `RecentList` (another candidate could have been `DroppingList` or `BoundedList` but I've thought that recent conveys the semantics better in the end). 

Because every insert `LPUSH` is followed by a `LTRIM`, we're practically guaranteed to always be at O(1) since we're only dropping an element at a time (and going at the end of the list is cheap obviously). 

## Test plan

Added unit tests.

### CODE REVIEW

**Comment by jhchabran** [comment]
nvm, doesn't fit what I needed.

File: internal/rcache/recent_list.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	key := q.globalPrefixKey()
+	res, err := redis.Values(c.Do("LRANGE", key, 0, -1))
+	if err != nil {
+		return nil, err
+	}
+	bs, err := redis.ByteSlices(res, nil)
+	if err != nil {
+		return nil, err
+	}
+	if len(bs) > q.size {
```
Comment (jhchabran) [inline internal/rcache/recent_list.go:74]: As we're not transacting in any way, it's possible when running concurrent inserts to end up with the list being bigger than it's maximal size. The slicing here ensures that we always return at most max elements.

File: internal/rcache/recent_list_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+				}
+			}
+			got, err := r.All(context.Background())
+			if err != nil {
+				t.Errorf("expected no error, got %q", err)
+			}
+			s, err := r.Size()
+			if err != nil {
+				t.Errorf("expected no error, got %q", err)
+			}
```
Comment (burmudar) [inline internal/rcache/recent_list_test.go:56]: Should we add a testcase for Slices too?

File: internal/rcache/recent_list.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// RecentList holds the most recently inserted items, discarding older ones if the total item count goes over the configured size.
+type RecentList struct {
+	key  string
+	size int
+}
+
+// NewRecentList returns a RecentCache, storing only a fixed amount of elements, discarding old ones if needed.
```
Comment (mrnugget) [inline internal/rcache/recent_list.go:18]:
```suggestion
// NewRecentList returns a RecentList, storing only a fixed amount of elements, discarding old ones if needed.
```

File: internal/rcache/recent_list.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"context"
+	"fmt"
+	"unicode/utf8"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// RecentList holds the most recently inserted items, discarding older ones if the total item count goes over the configured size.
```
Comment (mrnugget) [inline internal/rcache/recent_list.go:12]: Nitpick: I think "FIFO" (first in, first out) is more commonly used to describe the behavior here. "Recent" always makes me think in terms of dates and timestamps.

```suggestion
// FIFOList holds the most recently inserted items, discarding older ones if the total item count goes over the configured size.
```

File: internal/rcache/recent_list.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+// NewRecentList returns a RecentCache, storing only a fixed amount of elements, discarding old ones if needed.
+func NewRecentList(key string, size int) *RecentList {
+	return &RecentList{
+		key:  key,
+		size: size,
+	}
+}
+
+// Insert b in the cache and drops the last recently inserted item if the size exceeds the configured limit.
```
Comment (mrnugget) [inline internal/rcache/recent_list.go:26]:
```suggestion
// Insert b in the cache and drops the oldest inserted item if the size exceeds the configured limit.
```

File: internal/rcache/recent_list.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// All return all items stored in the RecentCache.
+//
+// This a O(n) operation, where n is the list size.
+func (l *RecentList) All(ctx context.Context) ([][]byte, error) {
+	return l.Slice(ctx, 0, -1)
+}
+
+// Slice return all items stored in the RecentCache between indexes from and to.
```
Comment (mrnugget) [inline internal/rcache/recent_list.go:69]:
```suggestion
// Slice return all items stored in the RecentList between indexes from and to.
```

File: internal/rcache/recent_list.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+	key := l.globalPrefixKey()
+	n, err := redis.Int(c.Do("LLEN", key))
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to execute redis command LLEN")
+	}
+	return n, nil
+}
+
+// All return all items stored in the RecentCache.
```
Comment (mrnugget) [inline internal/rcache/recent_list.go:62]:
```suggestion
// All return all items stored in the RecentList.
```

**Review by mrnugget — APPROVED** [review]

File: internal/rcache/recent_list.go (hunk 4 above)
Comment (jhchabran) [reply internal/rcache/recent_list.go:12]: Makes sense, applying

**Pull request merged**